daemon/cluster/controllers/plugin/controller.go

```diff
@@ -1,79 +1,264 @@
+	Pull(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer, opts ...plugin.CreateOpt) error
+	Upgrade(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer) error
+	Get(name string) (*v2.Plugin, error)
+	SubscribeEvents(buffer int, events ...plugin.Event) (eventCh <-chan interface{}, cancel func())
+}
 
 // NewController returns a new cluster plugin controller
-func NewController() (*Controller, error) {
-	return &Controller{}, nil
+func NewController(backend Backend, t *api.Task) (*Controller, error) {
+	payload, err := readPayload(t)
+	if err != nil {
+		return nil, err
+	}
+	return &Controller{backend: backend, config: payload, serviceID: t.ServiceID, logger: logrus.WithFields(logrus.Fields{
+		"controller": "plugin",
+		"task":       t.ID,
+		"plugin":     payload.Spec.Name,
+	})}, nil
 }
 
 // Update is the update phase from swarmkit
 func (p *Controller) Update(ctx context.Context, t *api.Task) error {
-	logrus.WithFields(logrus.Fields{
-		"controller": "plugin",
-	}).Debug("Update")
+	p.logger.Debug("Update")
 	return nil
 }
 
+func readPayload(t *api.Task) (runtime.GenericPluginPayload, error) {
+	var cfg runtime.GenericPluginPayload
+
+	generic := t.Spec.GetGeneric()
+	if err := proto.Unmarshal(generic.Payload.Value, &cfg); err != nil {
+		return cfg, errors.Wrap(err, "error reading plugin config")
+	}
+	return cfg, nil
+}
+
 // Prepare is the prepare phase from swarmkit
-func (p *Controller) Prepare(ctx context.Context) error {
-	logrus.WithFields(logrus.Fields{
-		"controller": "plugin",
-	}).Debug("Prepare")
+func (p *Controller) Prepare(ctx context.Context) (err error) {
+	p.logger.Debug("Prepare")
+
+	remote, err := reference.ParseNormalizedNamed(p.config.Spec.Remote)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing remote reference %q", p.config.Spec.Remote)
+	}
+
+	if p.config.Spec.Name == "" {
+		p.config.Spec.Name = remote.String()
+	}
+
+	// retrieve auth config from encoded auth
+	var authConfig enginetypes.AuthConfig
+	if p.config.PullOptions.RegistryAuth != "" {
+		if err := json.NewDecoder(base64.NewDecoder(base64.URLEncoding, strings.NewReader(p.config.PullOptions.RegistryAuth))).Decode(&authConfig); err != nil {
+			p.logger.Warnf("invalid auth config: %v", err)
+		}
+	}
+
+	privs := convertPrivileges(p.config.Spec.Privileges)
+
+	pl, err := p.backend.Get(p.config.Spec.Name)
+
+	defer func() {
+		if pl != nil && err == nil {
+			pl.Acquire()
+		}
+	}()
+
+	if err == nil && pl != nil {
+		if pl.SwarmServiceID != p.serviceID {
+			return errors.Errorf("plugin already exists: %s", p.config.Spec.Name)
+		}
+		if pl.IsEnabled() {
+			if err := p.backend.Disable(pl.GetID(), &enginetypes.PluginDisableConfig{ForceDisable: false}); err != nil {
```

Comment:
I guess we can check the image digest to see if it's actually changed here (and maybe the plugin backend should handle it and just return nil if there's nothing to do).